chrome/content/zotero/xpcom/translation/translate_item.js

```diff
@@ -857,19 +858,35 @@ Zotero.Translate.ItemSaver.prototype = {
 			});
 		}
 		
-		// Import from URL
 		let mimeType = attachment.mimeType ? attachment.mimeType : null;
 		let fileBaseName;
 		if (parentItemID) {
 			let parentItem = yield Zotero.Items.getAsync(parentItemID);
 			fileBaseName = Zotero.Attachments.getFileBaseNameFromItem(parentItem);
 		}
 		
-		Zotero.debug('Importing attachment from URL');
 		attachment.linkMode = "imported_url";
 		
 		attachmentCallback(attachment, 0);
-		
+
+		// Import from SingleFileZ Page Data
+		if (attachment.pageData && this._pageData) {
```

Comment:
Why both?

Reply:
No particular reason. Just a habit of checking to try and prevent errors. `attachment.pageData` is simply a boolean indicating the attachment should have pageData. `this._pageData` is the global object holding the pageData itself. If any attachment has pageData the global object should exist.

Reply:
But is there any reason for the global object, as opposed to just passing the data with the attachment it pertains to?

Reply:
While it wouldn't be the case now if we're only running SingleFile on the current page in the connector, couldn't there in theory be multiple attachments with page data?

Reply:
The reason I had the global object was because I was accounting for the scenario where the connector saves multiple items with the same page data. For example [https://ui.adsabs.harvard.edu/search/p_=0&q=test&sort=date%20desc%2C%20bibcode%20desc](this page) saves multiples items with a snapshot of the search text and a pdf from the result. Passing 10 copies of the snapshot data seemed to be compounding the problem of size so I passed one as a global.

SingleFile for multiple attachments only makes sense if we are loading those in the browser somehow to retrieve the page contents and I thought we would handle that in the client with a hidden browser. That may very well be the wrong approach though!

Reply:
Oh, yeah, that's just bad behavior on the part of the translator — it uses the same scrape() function for both single and multiple results, and it's passing the parent doc to every item, which makes sense for single items but not for multiple. So we should just fix that translator, and we definitely don't need to optimize for that here. Better to just keep this scoped to the attachment.

Reply:
Ok that makes sense. Thanks! I will switch that back now.